// react-component-grid/lib/Cell.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function cellData(data, x, y) {
    if (!data) {
        return null;
    }
    if (y >= data.length) {
        return null;
    }
    if (x >= data[y].length) {
        return null;
    }
    return data[y][x];
}

/**
 * @component
 *
 * `Cell`s are created by `Grid`s. They each have a single child, a clone of the component passed to the `component` prop of `Grid`.
 * `Cell`s are responsible for x / y positioning, and passing cell-specific props down to your cell components.
 *
 * @childprop {*} data The data for this cell from `Grid`s `data` prop.
 * @childprop {number} width The width of the cell, in pixels.
 * @childprop {number} height The height of the cell, in pixels.
 * @childprop {number} x The x index of this cell.
 * @childprop {number} y The y index of this cell.
 * @childprop {boolean} top Is true if the cell is on the top row.
 * @childprop {boolean} bottom Is true if the cell is on the bottom row.
 * @childprop {boolean} left Is true if the cell is on the left-most column.
 * @childprop {boolean} right Is true if the cell is on the right-most column.
 * @childprop {boolean} hover Is true if the cursor is currently hovering over this cell.
 * @childprop {number|null} hoverHitArea The index of the hit area that is currently being hovered, or `null` if the cursor is not hovering over this cell.
 */

function Cell(props) {
    var Component = props.component,
        data = props.data,
        x = props.x,
        y = props.y,
        cols = props.cols,
        rows = props.rows,
        width = props.width,
        height = props.height,
        hover = props.hover,
        className = props.className,
        additionalProps = props.additionalProps;


    var style = {
        left: x * width,
        top: y * height,
        display: 'block',
        position: 'absolute'
    };

    var isHovered = hover && hover.x == x && hover.y == y;

    return _react2.default.createElement(
        'div',
        {
            style: style,
            className: className
        },
        _react2.default.createElement(Component, _extends({}, additionalProps, {
            data: cellData(data, x, y),
            width: width,
            height: height,
            x: x,
            y: y,
            top: y == 0,
            bottom: y == rows - 1,
            left: x == 0,
            right: x == cols - 1,
            hover: isHovered,
            hoverHitArea: isHovered ? hover.h : null
        }))
    );
}

Cell.propTypes = {
    component: _react.PropTypes.oneOfType([_react.PropTypes.object, _react.PropTypes.func]).isRequired,
    x: _react.PropTypes.number.isRequired,
    y: _react.PropTypes.number.isRequired,
    cols: _react.PropTypes.number.isRequired,
    rows: _react.PropTypes.number.isRequired,
    width: _react.PropTypes.number.isRequired,
    height: _react.PropTypes.number.isRequired,
    hover: _react.PropTypes.shape({
        x: _react.PropTypes.number,
        y: _react.PropTypes.number,
        h: _react.PropTypes.number
    }),
    data: _react.PropTypes.any,
    additionalProps: _react.PropTypes.object
};

exports.default = Cell;
